File: Algorithms/Sorting/QuickSort/python/QuickSort.py
import sys
sys.setrecursionlimit(1500)

def partition(arr, start, end):
    pivot = arr[end]
    p_index = start
    for i in range(start, end):
        if arr[i] <= pivot:
            temp = arr[i]
            arr[i] = arr[p_index]
            arr[p_index] = temp
            p_index += 1
    temp = arr[p_index]
    arr[p_index] = arr[end]
    arr[end] = temp
    return p_index

def quick_sort(arr, start, end):
    if start < end:
        p_index = partition(arr, start, end)
        quick_sort(arr, start, p_index-1)
        quick_sort(arr, p_index+1, end)


if __name__ == '__main__':
    arr = list(map(int, input('Enter the array: ').split()))
    print("Array is: " + ' '.join([str(i) for i in arr]))
    print('Sorting...')
    quick_sort(arr, 0, len(arr)-1)
    print("Array is: " + ' '.join([str(i) for i in arr]))

File: Algorithms/Sorting/QuickSort/java/QuickSort.java
import java.util.Scanner;

public class QuickSort {

    public static void input(int[] arr, int size) {
        System.out.println("Enter the array: ");
        Scanner in = new Scanner(System.in);
        for(int i=0; i<size; i++)
            arr[i] = in.nextInt();
        in.close();
    }

    public static int partition(int[] arr, int start, int end) {
        int pivot = arr[end];
        int p_index = start;
        int temp;
        for(int i=start; i<end; i++) {
            if(arr[i] <= pivot) {
                temp = arr[i];
                arr[i] = arr[p_index];
                arr[p_index] = temp;
                p_index++;
            }
        }
        temp = arr[p_index];
        arr[p_index] = arr[end];
        arr[end] = temp;
        return p_index;
    }

    public static void quickSort(int[] arr, int start, int end) {
        if(start < end) {
            int p_index = partition(arr, start, end);
            quickSort(arr, start, p_index-1);
            quickSort(arr, p_index+1, end);
        }
    }

    public static void printArray(int[] arr, int size) {
        System.out.print("Array is: ");
        for(int i=0; i<size; i++)
            System.out.print(arr[i] + " ");
        System.out.println();
    }
}
